### PR TITLE
*: add builtin string function SPACE

### PR DESCRIPTION
**Pull request by tiancaiamao**
https://github.com/pingcap/tidb/issues/310

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -2647,6 +2648,10 @@ FunctionCallNonKeyword:
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string)), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
 	}
+| "SPACE" '(' Expression ')'
```
Comment (shenli) [inline parser/parser.y]: Seems this is not aligned. We use tab which is 8 spaces width.

File: parser/scanner.l [hunk 2]
```diff
@@ -972,6 +973,8 @@ redundant		lval.item = string(l.val)
 			return session
 {some}			lval.item = string(l.val)
 			return some
+{space}			lval.item = string(l.val)
```
Comment (shenli) [inline parser/scanner.l]: Keep alignment.

File: evaluator/builtin_string.go [hunk 3]
```diff
@@ -186,6 +187,26 @@ func builtinReverse(args []types.Datum, _ context.Context) (d types.Datum, err e
...
 }
 
+// See: http://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_space
+func builtinSpace(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+	x := args[0]
+	switch x.Kind() {
+	case types.KindNull:
+		return d, nil
+	default:
+		v, err := x.ToInt64()
```
Comment (shenli) [inline evaluator/builtin_string.go]: It is better to return error when err != nil. We use strict mode which will treat warning as error.
mysql> select space("abc");
+--------------+
| space("abc") |
+--------------+
|              |
+--------------+
1 row in set, 2 warnings (0.01 sec)

File: evaluator/builtin_string.go [hunk 4]
```diff
@@ -186,6 +187,26 @@ func builtinReverse(args []types.Datum, _ context.Context) (d types.Datum, err e
 	}
 }
 
+// See: http://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_space
+func builtinSpace(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+	x := args[0]
+	switch x.Kind() {
```
Comment (shenli) [inline evaluator/builtin_string.go]: if x.IsNull() {
    return d, nil
}
Then you do not need the long default part.

File: parser/scanner.l (hunk 2 above)
Comment (coocood) [inline parser/scanner.l]: It is aligned by tab, as {start} line below

File: evaluator/builtin_string.go [hunk 5]
```diff
@@ -186,6 +187,26 @@ func builtinReverse(args []types.Datum, _ context.Context) (d types.Datum, err e
...
+	case types.KindNull:
+		return d, nil
+	default:
+		v, err := x.ToInt64()
+		if err != nil || v < 0 {
+			v = 0
+		}
+		if v > math.MaxInt32 {
```
Comment (coocood) [inline evaluator/builtin_string.go]: MySQL returns null when v > 4*1024*1024

Comment (shenli) [inline evaluator/builtin_string.go]: It is more complicated, it depends on max_allowed_packet. We can keep it simple. Just set a threshold value.
  if ((ulonglong) count > INT_MAX32)
    count= INT_MAX32;

  // Safe length check
  tot_length= (uint) count * cs->mbminlen;
  if (tot_length > current_thd->variables.max_allowed_packet)
  {
    push_warning_printf(current_thd, Sql_condition::SL_WARNING,
                        ER_WARN_ALLOWED_PACKET_OVERFLOWED,
                        ER(ER_WARN_ALLOWED_PACKET_OVERFLOWED),
                        func_name(),
                        current_thd->variables.max_allowed_packet);
    goto err;
   }

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**